File: tree/543.diameter-of-binary-tree.cpp
// Created by Eric on 2020/3/11.

#include <iostream>
#include <algorithm>
#include <stack>
#include <queue>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// 树的层次遍历
void PrintTree(TreeNode *root) {
    TreeNode *temp = root;
    queue<TreeNode *> q;
    q.push(temp);

    while (!q.empty()) {
        TreeNode *cur = q.front();
        q.pop();

        cout << cur->val << " ";
        if (cur->left) {
            q.push(cur->left);
        }
        if (cur->right) {
            q.push(cur->right);
        }
    }
    cout << endl;
}

// HINT
// 当前节点的最大值等于左子树的最高层数+右子树的最高层树.
class Solution {
public:
    int res = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        maxdiameter(root);
        return res;
    }

    int maxdiameter(TreeNode *root){
        if (root == nullptr){
            return 0;
        }

        int lheight = maxdiameter(root->left);
        int rheight = maxdiameter(root->right);
        res = max(res, lheight + rheight);

        return max(lheight, rheight)+ 1;
    }
};

int main(){
    auto *l11 = new TreeNode(1);
    auto *l12 = new TreeNode(2);
    auto *l13 = new TreeNode(3);
    auto *l14 = new TreeNode(4);
    auto *l15 = new TreeNode(5);

    l11->left= l12;
//    l11->right = l13;
//    l12->left = l14;
//    l12->right = l15;

    Solution solution;
    int res = solution.diameterOfBinaryTree(l11);
    cout << res << endl;
}

File: tree/543.diameter-of-binary-tree.py
# -*- coding:utf-8 -*-

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.ans = 0

        def helper(root):
            if root is None: return 0

            left = helper(root.left)
            right = helper(root.right)
            self.ans = max(self.ans, left + right)
            # print(left, right)
            return max(left, right) + 1

        helper(root)
        return self.ans


if __name__ == '__main__':
    t1 = TreeNode(1)
    t2 = TreeNode(2)
    t3 = TreeNode(3)
    t4 = TreeNode(4)
    t5 = TreeNode(5)
    t1.left = t2
    t1.right = t3
    t2.left = t4
    t2.right = t5
    ans = Solution().diameterOfBinaryTree(t1)
    print('ans: ', ans)
